app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java

```diff
@@ -794,39 +476,589 @@ public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteris
+        final long poll_time = Math.max((Constants.SECOND_IN_MS * 5) + poll_backoff, POLLING_PERIOD - JoH.msSince(lastPacketTime));
+        if (poll_backoff < (Constants.MINUTE_IN_MS * 6)) {
+            poll_backoff += Constants.SECOND_IN_MS;
+        }
+        Log.d(TAG, "Scheduling next poll in: " + JoH.niceTimeScalar(poll_time) + " @ " + JoH.dateTimeText(poll_time + JoH.tsl()) + " period diff: " + (POLLING_PERIOD - JoH.msSince(lastPacketTime)));
+        return poll_time;
+    }
+
+    synchronized void checkConnection() {
+        status("Attempting connection");
+        final BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(Context.BLUETOOTH_SERVICE);
+        if (bluetoothManager == null) {
+            status("No bluetooth manager");
+            setRetryTimer();
+            return;
+        }
+
+        mBluetoothAdapter = bluetoothManager.getAdapter();
+        if (mBluetoothAdapter == null) {
+            status("No bluetooth adapter");
+            setRetryTimer();
+            return;
+        }
+
+        if (!mBluetoothAdapter.isEnabled()) {
+            if (Pref.getBoolean("automatically_turn_bluetooth_on", true)) {
+                Log.i(TAG, "Turning bluetooth on as appears disabled");
+                status("Turning bluetooth on");
+                JoH.setBluetoothEnabled(getApplicationContext(), true);
+            } else {
+                Log.d(TAG, "Not automatically turning on bluetooth due to preferences");
+            }
+        }
+
+        if (device != null) {
+            //mConnectionState = STATE_DISCONNECTED;
+            for (BluetoothDevice bluetoothDevice : bluetoothManager.getConnectedDevices(BluetoothProfile.GATT)) {
+                if (bluetoothDevice.getAddress().compareTo(device.getAddress()) == 0) {
+                    mConnectionState = STATE_CONNECTED;
+                    break;
+                }
+            }
+        }
+
+        Log.i(TAG, "checkConnection: Connection state: " + getStateStr(mConnectionState));
+        if (mConnectionState == STATE_DISCONNECTED || mConnectionState == STATE_DISCONNECTING) {
+            final ActiveBluetoothDevice btDevice = ActiveBluetoothDevice.first();
+            if (btDevice != null) {
+                final String deviceAddress = btDevice.address;
+                mDeviceAddress = deviceAddress;
+                try {
+                    if (mBluetoothAdapter.isEnabled() && mBluetoothAdapter.getRemoteDevice(deviceAddress) != null) {
+                        status("Connecting" + (Home.get_engineering_mode() ? ": " + deviceAddress : ""));
+                        connect(deviceAddress);
+                        mStaticState = mConnectionState;
+                        return;
+                    }
+                } catch (IllegalArgumentException e) {
+                    Log.e(TAG, "IllegalArgumentException: " + e);
+                }
+            }
+        } else if (mConnectionState == STATE_CONNECTING) {
```

Comment:
Here we manually manage the connecting state and shut it down if we think its been trying for too long instead of letting the framework manage this aspect.